### ISSUE TITLE
findFilesWithPattern: wrong search pattern as input

### ISSUE THREAD

**Issue by infeo** [cryptomator/dokany-nio-adapter#19]
Dokany-nio-adapter version: c3c91de2626010534e417ed3472e808ffa89f265
Dokany version: 1.2.1.2000
OS: Windows 10 Pro 1809
Related bug: [cryptomator/cryptomator#837](https://github.com/cryptomator/cryptomator/issues/837)

## Description
Some search patterns are translated from the windows world to the dokany-nio-adapter ~~wrongly~~ and thus in the current version there will be no or the wrong search results shown in a directory listing.

## Steps to reproduce:
1. Start the ReadWriteMirrorExample with an Folder containing a file named `test.txt`
2. Open a terminal
3. Change to the mounted directory and enter `dir test.*`

## Results
**Expected:** `test.txt` appears in the file listing
**Actual:** No file is found

## Log
```
11:06:06:036 [Thread-2314] TRACE org.cryptomator.frontend.dokany.ReadWriteAdapter - Try to open M:\test as Directory.
11:06:06:036 [Thread-2314] TRACE org.cryptomator.frontend.dokany.ReadWriteAdapter - (340) M:\test opened successful with handle 340.
11:06:06:037 [Thread-2315] TRACE org.cryptomator.frontend.dokany.ReadWriteAdapter - (340) getFileInformation() is called for M:\test.
11:06:06:037 [Thread-2315] TRACE org.cryptomator.frontend.dokany.ReadWriteAdapter - (340) File Information successful read from M:\test.
11:06:06:038 [Thread-2316] TRACE org.cryptomator.frontend.dokany.ReadWriteAdapter - (340) getFileInformation() is called for M:\test.
11:06:06:038 [Thread-2316] TRACE org.cryptomator.frontend.dokany.ReadWriteAdapter - (340) File Information successful read from M:\test.
11:06:06:039 [Thread-2317] TRACE org.cryptomator.frontend.dokany.ReadWriteAdapter - (340) findFilesWithPattern() is called for M:\test with search pattern test"*.
11:06:06:039 [Thread-2317] TRACE org.cryptomator.frontend.dokany.ReadWriteAdapter - (340) Successful searched content in M:\test.
11:06:06:039 [Thread-2318] TRACE org.cryptomator.frontend.dokany.ReadWriteAdapter - (340) cleanup() is called for M:\test.
11:06:06:040 [Thread-2319] TRACE org.cryptomator.frontend.dokany.ReadWriteAdapter - (340) closeFile() is called for M:\test.
```

**Comment by infeo** [comment]
depends on https://github.com/dokan-dev/dokany/issues/783

**Comment by infeo** [comment]
It seems like this is a MS-DOS compatibility feature and thus we have to account for that.

The rules i gatherd so far are the following:
1. `.` is replaced by `"`, if it is succeeded by `?` or `*` in the search string: `test.*` ↦ `test"*`
2. `*` is replaced by `<`, if it is succeed by `*` in the search string: `t*.txt` ↦ `t<.txt`
3. `?` is always replaced by `>`: `t?st.txt` ↦ `t>st.txt`

**Comment by Rondom** [comment]
I am not sure this fix does what you expect it to.

* [Java Globbing](https://docs.oracle.com/javase/10/docs/api/java/nio/file/FileSystem.html#getPathMatcher(java.lang.String)
) is different from Win32 globbing (which is different from [Windows-kernel-globbing](https://msdn.microsoft.com/en-us/library/ff546850(v=VS.85).aspx) patterns which [Dokan gets from the Kernel](https://dokan-dev.github.io/dokany-doc/html/group___dokan.html#ga2e8585b8730ee74c0cd0cadf3e8cb01a)).
* (irrelevant) [MS-DOS globbing](https://blogs.msdn.microsoft.com/oldnewthing/20071217-00/?p=24143) was even weirder than Win32-globbing, but the idiosyncratic nature of Win32-globbing is partly caused by backwards compatibility considerations
* The semantics of Win32-globbing should be completely irrelevant to your code if you are using the Java-API for globbing. So you need to do the conversion Windows-Kernel => Java and not Windows-Kernel => Win32.
* The string that Dokan gets from the kernel can contain any of the five characters, so I think your conversion is lossy and there will be multiple edge cases that don't work as one would expect them to.
* I remember that [ReactOS ](https://github.com/reactos/reactos/blob/8a4c5763dac0a3c8a0d3b776bd5d4529b3f7bf9b/modules/rostests/kmtests/ntos_fsrtl/FsRtlExpression.c)and Samba have some good test cases if I remember it correctly. You can use them to test your implementation. You will need to respect their licence though, so you probably cannot copy it (haven't checked what the license is).

Disclaimer: I haven't spend a long time to think of specific edge cases, but I think there should be some, because of the different semantics of the wildcard-characters between Windows-Kernel and Java and the way you do a simple replace.

**Comment by overheadhunter** [comment]
@Rondom I'm a bit confused by all those different dialects. [According to Dokany docs](https://dokan-dev.github.io/dokany-doc/html/struct_d_o_k_a_n___o_p_e_r_a_t_i_o_n_s.html#a0f280feb3cbff134e997bec6d6db5171) we get a "Windows MS-DOS-style expression". Is this the same MS-DOS globbing you've linked to under your second bullet point?

The ReactOS test cases are definitely a good point to start. [LGPLv2.1 is compatible with AGPLv3](https://www.gnu.org/licenses/gpl-faq.html#AllCompatibility).

**Comment by infeo** [comment]
@overheadhunter No. The globbing dokany (and thus we) gets is the first bullet aka windows kernel style globbing.

@Rondom Yes, you are correct. But i did'nt knew, that the symbols `<`,`>` and `"` represent the `DOS_DOT` and so on in the windows kernel. Since this is now clear (and even better in the documentation 😊 ), i feel able to fix it right.

**Comment by infeo** [comment]
After looking into the case we decided to drop the implementation of FindFilesWithPattern() and rely on the default implementation of dokany/windows which has native windows kernel pattern support.

Closed with 0435f23625d4f609c8a893d07ece257fad591623
